File: solutions/javascript/roman-numerals/1/roman-numerals.js
const NUMERALS = {
  1000: "M",
  900: "CM",
  500: "D",
  400: "CD",
  100: "C",
  90: "XC",
  50: "L",
  40: "XL",
  10: "X",
  9: "IX",
  5: "V",
  4: "IV",
  1: "I",
};

const FACTORS = [1000, 900, 500, 400, 100, 90, 50, 40, 10, 9, 5, 4, 1];

export const toRoman = (num) => {
  var output = "";

  FACTORS.forEach((factor) => {
    if (!(num < factor)) {
      var multiple = Math.floor(num / factor);
      var remainder = num % factor;

      output += NUMERALS[factor].repeat(multiple);

      num = remainder;
    }
  });

  return output;
};

File: solutions/ruby/roman-numerals/2/roman_numerals.rb
class Integer
  def to_roman
    romans = {
      1000 => "M",
      900 => "CM",
      500 => "D",
      400 => "CD",
      100 => "C",
      90 => "XC",
      50 => "L",
      40 => "XL",
      10 => "X",
      9 => "IX",
      5 => "V",
      4 => "IV",
      1 => "I"
    }
    int = self
    numeral = ""
    factors = romans.keys
    factors.each do |factor|
      next if int < factor
      multiple = int / factor
      remainder = int % factor
      numeral << romans[factor] * multiple
      int = remainder
    end
    numeral
  end
end

module BookKeeping
  VERSION = 2
end